agent/connect/ca/provider_consul.go

```diff
@@ -170,10 +170,126 @@ func (c *ConsulProvider) GenerateRoot() error {
 	return nil
 }
 
+// GenerateIntermediateCSR creates a private key and generates a CSR
+// for another datacenter's root to sign.
+func (c *ConsulProvider) GenerateIntermediateCSR() (string, error) {
+	_, providerState, err := c.getState()
+	if err != nil {
+		return "", err
+	}
+
+	if c.isRoot {
+		return "", fmt.Errorf("provider is the root certificate authority, " +
+			"cannot generate an intermediate CSR")
+	}
+
+	// Create a new private key and CSR.
+	signer, pk, err := connect.GeneratePrivateKey()
+	if err != nil {
+		return "", err
+	}
+
+	csr, err := connect.CreateCSR(c.spiffeID, signer)
```

Comment:
I think the CSR needs to specify that it wants a CA cert so that we can enforce that only CSRs that explicitly request a CA cert will have one signed by `SignIntermediate`. It seems sensible to me to at least not allow generic leaf CSRs to be signed as CA certs.

I _think_ so far this will just create a generic CSR with the signer SPIFFE URI SAN right?

Reply:
I looked into doing this and I don't know if we can enforce it since Vault doesn't set the IsCA constraint on the intermediate CSR that it generates (go's CertificateRequest type doesn't have the field like Certificate does, so you have to manually add it to the extensions in kind of a clunky way). We can still set it on the one the consul CA generates, but probably can't require it down in SignIntermediate.

Reply:
> since Vault doesn't set the IsCA constraint on the intermediate CSR that it generates

Hmm OK that's a fair reason so skip it for now. It's not technically necessary but seemed like a nice way to tighten up constraints. But we should probably keep all providers compatible here so that you could for example sign a Vault-generated CSR (with no IsCA) with a built-in root to make it simpler to reason about migrations and different orders of bootstrapping etc. Of course it's most reasonable that you use the same CA provider for all datacenters but feels like we shouldn't make that decision only based on artificial CSR compatibility.

So yeah maybe we can generate CSR with one for the sake of completness but not require it just yet.

So I think this is good to go.